### DISCUSSION TITLE
npm run build keeps lots of jest-worker child process and eating up all memories after some builds

### DISCUSSION THREAD

**Discussion by hkmsadek** [vercel/next.js#42668] [Help]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

 Operating System:
      Platform: linux
      Arch: x64
      Version: #58-Ubuntu SMP Thu Oct 13 08:03:55 UTC 2022
    Binaries:
      Node: 16.17.0
      npm: 8.15.0
      Yarn: 1.22.19
      pnpm: N/A
    Relevant packages:
      next: 13.0.3-canary.1
      eslint-config-next: N/A
      react: 18.2.0
      react-dom: 18.2.0

### Describe the Bug

We have been running some npm run builds commands as we are publishing updates 5/6 times a day, we noticed that
at, npm run build leaves lot of jest-worker child process instances like the below image. 
![Uploading Screenshot 2022-11-09 at 12.25.12 am.png…]()

It's eating all the memory and getting locked. 

### Expected Behavior

It shouldn't eat lot of memories and child instances should get deleted after build? 

### Link to reproduction

it's a production app not sure how to add

### To Reproduce

just running npm run build multiple times in a large projects will leave this issue

**Comment by hkmsadek** [comment]
![Uploading Screenshot 2022-11-09 at 12.25.12 am.png…]()

**Comment by icyJoseph** [comment]
Do you run tests as part of your build process? Do you observe the same by just running `npx next build`? Also your screenshots are not visible.

**Comment by mohas** [comment]
We also have this issue, we do not have any test runs in our deployment runs

**Reply by icyJoseph** [reply]
Do you have any screenshots, logs, or repro demo?

**Reply by mohas** [reply]
Of course, we will work something out when I get to the office tomorrow

**Reply by mohas** [reply]
hey @icyJoseph  look at the sheer number of these processes, there are literally hundreds of them

![image](https://user-images.githubusercontent.com/1816728/201517824-f697afbe-8816-4517-a58f-d68fc9a67ebd.png)

**Reply by icyJoseph** [reply]
Ok, out of the box trying this on my local machine with a dummy project does nothing. You do have tests anyway right?

## Edit

With node 16.13.0, and jest tests running, I still see no problem. 

Could it be that you are importing from a test file into a component/page?

**Reply by mohas** [reply]
we do have tests, both in __test__ directory and alongside js files, I don't think there is a case of importing from a test file, does the jest process run as a part of build process at all? and if it is can we disable running tests? so maybe we can disable testing and build again to see what happens.
do you need any more detail from the processes?

**Reply by icyJoseph** [reply]
AFAIK, then this might be something related to the actual Workers used by Next.js to build pages in parallel. So this has nothing to do with tests, this is just a product made by the jest team to run work in parallel. 

What next.js version are you running? Are you using any special next.config.js experimentals?

**Reply by icyJoseph** [reply]
With:

```js
const nextConfig = {
  reactStrictMode: true,
  swcMinify: true,
  experimental: {
    workerThreads: true,
    cpus: 4,
    sharedPool: true,
  },
};
```

And building 1000 pages on build I still can't get this issue, my CPU and RAM go back down to regular levels after build and there's no jest-worker processes running. 

I am in MacOS though... could you try to provide a minimal reproduction demo?

**Reply by mohas** [reply]
we are using Next@12.2.5, btw what does sharedPool do in this config?

**Reply by icyJoseph** [reply]
I still can't trigger this 😞 Wonder if it is OS related. I don't know about these experimental flags, I just started to play with them, given that they affect workers at build time, at least from inspecting the code it seems that way.

**Reply by mohas** [reply]
Ok thank you, I will try some more to see what may trigger these, will post more if I see something

**Comment by IronDdtRIiGh** [comment]
I run it  s subserver issntsi.u.c.e.m- abcdefghighlmnopqrstuvwxy.!
Apptsdinue. e. u*^+

**Comment by brandonfake69** [comment]
Running into this issue as well. Crashes my server. Tried adding the following to my docker-compose file with no luck:

```
  NODE_OPTIONS: --max-old-space-size=512
  NEXT_TELEMETRY_DISABLED: 1
```

**Reply by brandonfake69** [reply]
Also tried adding some optimizations to `next.config.ts` (also didn't work):

```
const nextConfig: NextConfig = {
  // Reduce memory usage during build
  experimental: {
    workerThreads: false,
    cpus: 1,
    optimizePackageImports: ["@google-cloud/kms", "twilio"],
  },

  // Disable source maps in production to save memory
  productionBrowserSourceMaps: false,

  // Reduce bundle size
  swcMinify: true,

  // Limit webpack memory usage
  webpack: (config, { isServer }) => {
    if (!isServer) {
      config.optimization = {
        ...config.optimization,
        splitChunks: {
          chunks: "all",
          cacheGroups: {
            vendor: {
              test: /[\\/]node_modules[\\/]/,
              name: "vendors",
              chunks: "all",
            },
          },
        },
      };
    }

    // Limit memory usage
    config.optimization = {
      ...config.optimization,
      moduleIds: "deterministic",
    };

    return config;
  },
};
```
